filament/backend/src/opengl/OpenGLDriver.cpp

```diff
@@ -635,56 +629,65 @@ void OpenGLDriver::createTextureSwizzledR(Handle<HwTexture> th,
-            t->gl.targetIndex = (uint8_t)gl.getIndexForTextureTarget(GL_TEXTURE_CUBE_MAP);
-            break;
-    }
+    if (UTILS_LIKELY(usage & TextureUsage::SAMPLEABLE)) {
+        // We DO NOT update targetIndex at function exit to take advantage of the fact that
+        // getIndexForTextureTarget() is constexpr -- so all of this disappears at compile time.
+        switch (target) {
+            case SamplerType::SAMPLER_EXTERNAL:
+                t->gl.target = GL_TEXTURE_EXTERNAL_OES;
+                t->gl.targetIndex = (uint8_t)gl.getIndexForTextureTarget(GL_TEXTURE_EXTERNAL_OES);
+                break;
+            case SamplerType::SAMPLER_2D:
+                t->gl.target = GL_TEXTURE_2D;
+                t->gl.targetIndex = (uint8_t)gl.getIndexForTextureTarget(GL_TEXTURE_2D);
+                break;
+            case SamplerType::SAMPLER_3D:
+                t->gl.target = GL_TEXTURE_3D;
+                t->gl.targetIndex = (uint8_t)gl.getIndexForTextureTarget(GL_TEXTURE_3D);
+                break;
+            case SamplerType::SAMPLER_2D_ARRAY:
+                t->gl.target = GL_TEXTURE_2D_ARRAY;
+                t->gl.targetIndex = (uint8_t)gl.getIndexForTextureTarget(GL_TEXTURE_2D_ARRAY);
+                break;
+            case SamplerType::SAMPLER_CUBEMAP:
+                t->gl.target = GL_TEXTURE_CUBE_MAP;
+                t->gl.targetIndex = (uint8_t)gl.getIndexForTextureTarget(GL_TEXTURE_CUBE_MAP);
+                break;
+        }
 
-    if (t->samples > 1) {
-        // Note: we can't be here in practice because filament's user API doesn't
-        // allow the creation of multi-sampled textures.
-        if (gl.features.multisample_texture) {
-            // multi-sample texture on GL 3.2 / GLES 3.1 and above
-            t->gl.target = GL_TEXTURE_2D_MULTISAMPLE;
-            t->gl.targetIndex = (uint8_t)gl.getIndexForTextureTarget(GL_TEXTURE_2D_MULTISAMPLE);
-        } else {
-            // Turn off multi-sampling for that texture. It's just not supported.
+        if (t->samples > 1) {
+            // Note: we can't be here in practice because filament's user API doesn't
+            // allow the creation of multi-sampled textures.
+            if (gl.features.multisample_texture) {
+                // multi-sample texture on GL 3.2 / GLES 3.1 and above
+                t->gl.target = GL_TEXTURE_2D_MULTISAMPLE;
+                t->gl.targetIndex = (uint8_t)gl.getIndexForTextureTarget(GL_TEXTURE_2D_MULTISAMPLE);
+            } else {
+                // Turn off multi-sampling for that texture. It's just not supported.
+            }
         }
+    
+    } else {
+        assert_invariant(any(usage & (
+                TextureUsage::COLOR_ATTACHMENT |
+                TextureUsage::DEPTH_ATTACHMENT |
+                TextureUsage::STENCIL_ATTACHMENT)));
+        assert_invariant(levels == 1);
+        assert_invariant(target == SamplerType::SAMPLER_2D);
+        t->gl.target = GL_RENDERBUFFER;
```

Comment:
Would be worth a comment that these are the same assumptions and rules that exist in `OpenGLDriver::createTextureR()`: Sampleable -> GL texture, Non-sampleable -> GL renderbuffer, plus the asserts listed here.

Reply:
[8caf282](https://github.com/shapr3d/filament/pull/93/commits/8caf2825ff5752648410c6fb50de9081fda03b1d)